File: csharp/intermediate/010_Inheritance.cs
// [10] 상속과 다형성 - Inheritance & Polymorphism
// 레벨: 3 | C#의 상속과 다형성을 이해합니다

using System;

class Program
{
    static void Main()
    {
        Animal[] list =
        {
            new Dog("d",1),
            new Cat("c",2)
        };

        foreach (var a in list)
        {
            a.Speak();
        }

        ((Dog)list[0]).Run();
    }
}

abstract class Animal
{
    public string Name;
    public int Age;

    public Animal(string n,int a)
    {
        Name=n;
        Age=a;
    }

    public virtual void Speak()
    {
        Console.WriteLine("none");
    }
}

class Dog:Animal
{
    public Dog(string n,int a):base(n,a){}

    public override void Speak()
    {
        Console.WriteLine(Name+" bark");
    }

    public void Run()
    {
        Console.WriteLine(Name+" run");
    }
}

class Cat:Animal
{
    public Cat(string n,int a):base(n,a){}

    public override void Speak()
    {
        Console.WriteLine(Name+" meow");
    }
}

File: java/intermediate/010_Inheritance.java
// [10] 상속 (Inheritance) - Inheritance & Polymorphism
// 레벨: 3 | 클래스 상속과 메서드 오버라이딩을 학습합니다

public class Inheritance {

    public static void main(String[] args) {

        Animal a1 = new Dog("Rex");
        Animal a2 = new Cat("Milo");

        Animal[] list = {a1, a2};

        for (Animal a : list) {
            a.speak();
        }

        Dog d = new Dog("Max");
        d.run();
    }
}

class Animal {

    String name;

    Animal(String name) {
        this.name = name;
    }

    void speak() {
        System.out.println("sound");
    }
}

class Dog extends Animal {

    Dog(String name) {
        super(name);
    }

    void speak() {
        System.out.println(name + " bark");
    }

    void run() {
        System.out.println(name + " run");
    }
}

class Cat extends Animal {

    Cat(String name) {
        super(name);
    }

    void speak() {
        System.out.println(name + " meow");
    }
}
